src/metabase/sync/util.clj

```diff
@@ -424,7 +435,25 @@
    database :- i/DatabaseInstance
    sync-steps :- [StepDefinition]]
   (let [start-time    (t/zoned-date-time)
-        step-metadata (mapv #(run-step-with-metadata database %) sync-steps)
+        step-metadata (loop [[step-defn & rest-defns] sync-steps
```

Comment:
some minor quibbles: 
- instead of `step-defn` and `rest-defns`, maybe just `step` and `steps` or `step` and `remaining`. 
- no-more probably should not exist and just check `(seq steps)` or `(seq remaining)` where no-more is used. In general I'm pretty wary of negated variables and a common idiom is just `(seq steps)` or `(seq remaining)` whatever the renamed things become (if renamed).
- rename `should-not-cont` to `abandon?` and compute it with `(seq (set/intersection exception-classes-to-not-retry exception-class-names))`
- the nested ifs under the throwable branch can then become
```clojure
(cond abandon? new-result
      (not (seq remaining)) new-result
      :else (recur remaining new-result))
```
- Also another concern, is that in https://github.com/metabase/metabase/blob/master/src/metabase/sync.clj#L41 we run steps without checking whether previous steps failed. One thing we could do is mark the result `r` as unrecoverable if it is so and then this map above could be changed into a simple reduce or loop looking for any unrecoverable results and abandon, similar to what this does here. This is probably the source of what flamber is seeing as we'll do this dance once for each of those three steps.

Reply:
On the latter point,

> Please note that this function is *not* what is called by the scheduled tasks; those call different steps independently. This function is called when a Database is first added.

Can we handle that in another PR? I don't want to keep growing the scope of this, as the intention here was to unblock the scheduler and deal with scheduled syncs.

Reply:
good point. sorry i missed that